Mathlib/Data/Greedoid/Basic.lean

```diff
@@ -0,0 +1,152 @@
+/-
+Copyright (c) 2024 Jihoon Hyun. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jihoon Hyun
+-/
+import Mathlib.Data.Greedoid.Accessible
+import Mathlib.Data.Greedoid.Exchange
+
+/-!
+# Greedoid
+
+This file contains the definition of `ExchangeProperty` and `AccessibleProperty`, along with the
+main subject `Greedoid α`.
+
+Greedoid is a set system, i.e., a family of sets, over a finite ground set, which satisfies both
+exchange and accessible properties.
+
+Main definitions:
+
+* `Greedoid`: The greedoid structure.
+* `Greedoid.ground`: The ground set of the greedoid.
+* `Greedoid.Feasible`: The feasible property of the greedoid.
+* `Greedoid.Bases`: A bases of a greedoid `G` is a maximal feasible set of `G`.
+* `Greedoid.Basis`: A basis of a set `s` with respect to a greedoid `G` is a maximal subset of `s`
+  which is feasible in `G`.
+-/
+
+/-- Greedoid is a nonempty (finite) set system satisfying both accessible and exchange property. -/
+structure Greedoid (α : Type*) where
+  /-- The ground set which every element lies on. -/
+  ground : Finset α
+  /-- The feasible property of the greedoid. -/
+  Feasible : Finset α → Prop
+  feasible_empty : Feasible ∅
+  exchange_property : Greedoid.ExchangeProperty Feasible
+  subset_ground : ∀ s, Feasible s → s ⊆ ground
+
+section Greedoid
+
+variable {α : Type*}
+
+end Greedoid
+
+namespace Greedoid
+
+variable {α : Type*}
+
+open Nat List Finset
+
+theorem eq_of_veq : ∀ {G₁ G₂ : Greedoid α},
+    G₁.ground = G₂.ground → G₁.Feasible = G₂.Feasible → G₁ = G₂
+  | ⟨_, _, _, _, _⟩, ⟨_, _, _, _, _⟩, h₁, h₂ => by cases h₁; cases h₂; rfl
+
+@[simp]
+theorem feasible_injective :
+    Function.Injective (fun G : Greedoid α => (G.ground, G.Feasible)) :=
+  fun _ _ => by simp; exact eq_of_veq
+
+@[simp]
+theorem feasible_inj {G₁ G₂ : Greedoid α} :
+    G₁.ground = G₂.ground ∧ G₁.Feasible = G₂.Feasible ↔ G₁ = G₂ :=
+  ⟨fun h => by apply eq_of_veq <;> simp [h], fun h => by simp [h]⟩
+
+variable {G : Greedoid α}
+variable {s : Finset α}
```

Comment:
I personally would write `{s s₁ s₂ : Finset α}` together and leave out the explicit arguments from the `variable` declarations.